pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +96,458 @@ const (
+// rollout.
+func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
+	out := config.Out
+	oldRc := config.OldRc
+	scaleRetryParams := NewRetryParams(config.Interval, config.Timeout)
+
+	// Find an existing controller (for continuing an interrupted update) or
+	// create a new one if necessary.
+	sourceId := fmt.Sprintf("%s:%s", oldRc.Name, oldRc.UID)
+	newRc, existed, err := r.getOrCreateTargetController(config.NewRc, sourceId)
+	if err != nil {
+		return err
+	}
+	if existed {
+		fmt.Fprintf(out, "Continuing update with existing controller %s.\n", newRc.Name)
+	} else {
+		fmt.Fprintf(out, "Created %s\n", newRc.Name)
+	}
+	// Extract the desired replica count from the controller.
+	desired, err := strconv.Atoi(newRc.Annotations[desiredReplicasAnnotation])
+	if err != nil {
+		return fmt.Errorf("Unable to parse annotation for %s: %s=%s",
+			newRc.Name, desiredReplicasAnnotation, newRc.Annotations[desiredReplicasAnnotation])
+	}
+	// Extract the original replica count from the old controller, adding the
+	// annotation if it doesn't yet exist.
+	_, hasOriginalAnnotation := oldRc.Annotations[originalReplicasAnnotation]
+	if !hasOriginalAnnotation {
+		existing, err := r.c.ReplicationControllers(oldRc.Namespace).Get(oldRc.Name)
+		if err != nil {
+			return err
+		}
+		if existing.Annotations == nil {
+			existing.Annotations = map[string]string{}
+		}
+		existing.Annotations[originalReplicasAnnotation] = strconv.Itoa(existing.Spec.Replicas)
+		updated, err := r.c.ReplicationControllers(existing.Namespace).Update(existing)
+		if err != nil {
+			return err
+		}
+		oldRc = updated
+	}
+	original, err := strconv.Atoi(oldRc.Annotations[originalReplicasAnnotation])
+	if err != nil {
+		return fmt.Errorf("Unable to parse annotation for %s: %s=%s\n",
+			oldRc.Name, originalReplicasAnnotation, oldRc.Annotations[originalReplicasAnnotation])
+	}
+
+	// The maximum pods which can go unavailable during the update.
+	maxUnavailable, err := extractMaxValue(config.MaxUnavailable, "maxUnavailable", original)
+	if err != nil {
+		return err
+	}
+	// The maximum scaling increment.
+	maxSurge, err := extractMaxValue(config.MaxSurge, "maxSurge", original)
+	if err != nil {
+		return err
+	}
+	// Further validation.
+	if maxUnavailable == 0 && maxSurge == 0 {
+		return fmt.Errorf("one of maxSurge or maxUnavailable must be specified")
+	}
+	// The minumum pods which must remain available througout the update
+	// calculated for internal convenience.
+	minAvailable := original - maxUnavailable
+	// Set the initial increment. This will drive the direction of the update;
+	// an initial increment of 0 means the first scale-up will be skipped,
+	// causing a down/up direction.
+	increment := 0
+	// A surge value means we need to move up/down to make availability for a
+	// scale down, and we'll burst higher than if only a maxUnavailable were
+	// set.
+	if maxSurge > 0 {
+		increment = maxUnavailable + maxSurge
```

Comment:
On Fri, Aug 21, 2015 at 6:28 AM, Dan Mace <notifications@github.com> wrote:

> In pkg/kubectl/rolling_updater.go
> <https://github.com/kubernetes/kubernetes/pull/11942#discussion_r37633334>
> :
>
> > +	// Further validation.
> > +	if maxUnavailable == 0 && maxSurge == 0 {
> > +		return fmt.Errorf("one of maxSurge or maxUnavailable must be specified")
> > +	}
> > +	// The minumum pods which must remain available througout the update
> > +	// calculated for internal convenience.
> > +	minAvailable := original - maxUnavailable
> > +	// Set the initial increment. This will drive the direction of the update;
> > +	// an initial increment of 0 means the first scale-up will be skipped,
> > +	// causing a down/up direction.
> > +	increment := 0
> > +	// A surge value means we need to move up/down to make availability for a
> > +	// scale down, and we'll burst higher than if only a maxUnavailable were
> > +	// set.
> > +	if maxSurge > 0 {
> > +		increment = maxUnavailable + maxSurge
>
> No. I still think it should be (original + maxSurge) - (old + new).
> maxUnavailable has no role to play while deciding how much should I scale
> up.
>
> Maybe we should clarify the requirement for when maxUnavailable and
> maxSurge are >0. The requirements were outlined earlier by @bgrant0607
> <https://github.com/bgrant0607> in #12236 (comment)
> <https://github.com/kubernetes/kubernetes/pull/12236#discussion_r36554290>
> :
>
> maxUnavailabilityPercentage=30 and maxSurgePercentage=0: The size would be
> reduced by 30% immediately. New pods would be created as the old ones
> disappeared. Once the new pods were ready for long enough to be considered
> available, the old RC could be downsized further, followed by increasing
> the new RC more.
>
> maxUnavailabilityPercentage=0 and maxSurgePercentage = 30: The size of the
> new RC would be set to 30% of the original/desired immediately. The old RC
> would be downsized only after the new pods were ready for long enough to be
> considered available.
>
> maxUnavailabilityPercentage=10 and maxSurgePercentage = 20: The size of
> the new RC would be set to 30% of the original/desired immediately. The old
> RC would be downsized by 10% immediately. Progress thereafter would proceed
> as in the cases above.
>
> If we ignore maxUnavailable in the initial increment calculation, we don't
> implement example 3 as I understood it.
>
Yes. In this case, we should downsize the old RC by 10% before scaling up
the new RC to 30%.
If we scale up the new RC by 30% directly, then we run the risk of surging
by 30%, but the maxSurgePercentage is 20%.


> In the test case you added, change maxUnavailable to 2, and increment will
> turn out to be 3, while it should be 1.
>
> A maximum increment of 3 is what I'd expect here (surge+unavailable per
> example 3). Maybe walk through a couple more examples just to be sure I'm
> getting it?
>
If we increment by 3, then we will exceed the maxSurge specified by user.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/11942/files#r37633334>.
>